fs: Fix fs_open resource leak when invoked on fs_file_t object in use

de-nordic opened the pull request:
fs: Fix fs_open resource leak when invoked on fs_file_t object in use

Fixes problem when fs_open invoked on fs_file_t object, which is already
holding information on opened file, overwrites references to other
memory objects within the fs_file_t object causing resource leak.
If fs_open is invoked on already used fs_file_t object, it will return
-EBUSY.

Note: The change requires that all fs_file_t objects should be
initialized to 0.

Fixes: #29478

de-nordic committed:
fs: Fix fs_open resource leak when invoked on fs_file_t object in use

Fixes problem when fs_open invoked on fs_file_t object, which is already
holding information on opened file, overwrites references to other
memory objects within the fs_file_t object causing resource leak.
If fs_open is invoked on already used fs_file_t object, it will return
-EBUSY.

Note: The change requires that all fs_file_t objects should be
initialized to 0.

Fixes: #29478

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
settings: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/settings: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
samples: littlefs: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
samples: fs: Extend littlefs sample to native_posix and qemu_x86

Adds native_posix and qemu_x86 the platform_allow list for the
littlefs sample application.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests: fs: fatfs: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests: fs: littlefs: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests: fs: multifs: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu commented:
Maybe instead of `struct fs_file_t file = { 0 }` I would intoruce define of empty the fs_file_t variable or even introduce a macro which declares empty variable of fs_file_t type an use it everywhere.

de-nordic commented:
> 
> 
> Maybe instead of `struct fs_file_t file = { 0 }` I would introduce a define of the empty fs_file_t variable or even introduce a macro which declares empty variable of fs_file_t type and use it everywhere.

Something like:
```
FS_FILE_T_INIT(name) { name.mp = 0; }
// usage
fs_file_t some;
FS_FILE_T_INIT(some);
```
or
```
MK_FS_FILE_T(name) name = { 0 };
// usage
MK_FS_FILE_T(some);
```

pabigot reviewed (changes requested):
I'd like to see the commits that initialize the `fs_file_t` instances moved before the commit that requires they be initialized, otherwise bisect on fs tests may break.  Change the commit text to "in preparation for code changes that will require this" or some similar text instead of referring to `fs_open()` changes that won't be present yet.

I don't have a problem with `= { 0 };` as the initializer idiom, but I'd be fine with `FS_FILE_T_INITIALIZER` defined to `(struct fs_file_t){ 0 }` to make it more clear.  Or an inline function `fs_file_t_init()` that can be called.  Or both, with the latter using the former.

I **really** don't like macro functions to define initialized instances (among other things, they trigger checkpatch errors because it looks like code after declarations without an empty line), but since others in the community do like them I wouldn't block on it.

In any case the fact that you now need to initialize instances of `struct fs_file_t` before their first use must be part of the documentation of `struct fs_file_t`.

de-nordic commented:
>  an inline function `fs_file_t_init()` that can be called

I will create the function. Now, as I think about it, the `= { 0 };` may require new changes later, when the `fs_type_f` gets extended with fields that require non-0 initialization, if such thing would happen
Not that I am not willing to do the changes again later, just users may not be happy to go through their code again.

de-nordic committed:
fs: Add fs_file_t_init() function

The fs_file_t_init() function has been added that should be used
for initialization of fs_file_t structures before passing them
to fs_open and other functions.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
posix: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
fs: shell: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
native_posix: fuse: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

pabigot commented on include/fs/fs.h:
```diff
@@ -220,6 +220,22 @@ struct fs_statvfs {
+ */
+inline void fs_file_t_init(struct fs_file_t *zfp)
+{
+	zfp->mp = NULL;
```
Would you consider:

    *fp = (struct fs_file_t){
      flags = 0,
    };

This will initialize all fields to zero.  The only reason you need to provide even one is just because it's required syntactically.

This is a bit more flexible in that if you add or rename fields that would be initialized to zero you don't need to touch this function: they'll automatically be zeroed, which wouldn't happen if they're all set individually.

de-nordic replied:
@pabigot I have done that before, but this will fail compilation wit
```
tests/subsys/fs/fat_fs_dual_drive/src/main.c:18: undefined reference to `fs_file_t_init
```
and many other places, unless I will define the function as `static inline`.

pabigot replied:
I hadn't noticed: why is this *not* `static inline`?  A non-static inline will fail exactly in that way unless you add this to some implementation file:

```
extern void fs_file_t_init(struct fs_file_t *zfp);
```

This guarantees that there is an extern version available for modules where the inline did not occur.  It comes from [C99](https://en.cppreference.com/w/c/language/inline), and for whatever reason we don't generally use it in Zephyr, preferring to use the old GNU extension `static inline` idiom which forces the inline (or a local callable implementation.

So, you can make this `static inline` so it's like almost every existing header inline function in Zephyr, or add the `extern` declaration to an implementation file.

de-nordic replied:
Still: why without static the code with separate assignments linked fine while the one with cast did not?
Doesn't that mean that the assignments were always inlined while the other code could not be inlined in some situations?

pabigot replied:
No idea.  The compiler's free to inline or not based on whatever rules it's using.

pabigot commented on subsys/fs/fuse_fs_access.c:
```diff
@@ -468,6 +469,12 @@ static void fuse_fs_access_init(void)
 {
 	int err;
 	struct stat st;
+	int i = 0;
```
Better to have this be `size_t` since it can never be negative and it's being compared to a value of type `size_t`.

pabigot commented on tests/subsys/fs/fs_api/src/main.c:
```diff
@@ -72,6 +72,7 @@ void test_main(void)
 			 ztest_unit_test_setup_teardown(test_mount,
 							fs_setup,
 							dummy_teardown),
+			 ztest_unit_test(test_fs_file_t_init),
```
Why isn't this in the commit that adds this test case?

de-nordic replied:
Some accidental addition to commit amend.

pabigot commented on tests/subsys/fs/fs_api/src/test_fs.h:
```diff
@@ -32,6 +32,7 @@ struct test_fs_data {
 	int reserve;
 };
 
+void test_fs_file_t_init(void);
```
this too.

pabigot reviewed (dismissed):
The addition of `test_fs_init_t` seems to be split across two commits; it should all be together in the commit that adds the test.  This is the blocking request.

The other comments aren't blockers, just suggestions.

de-nordic committed:
tests: lib: gui: lvgl: Add fs_file_t type variable initializations

The commit adds initializations of fs_file_t variables in preparation
for fs_open function change that will require fs_file_t object, passed
to the function, to be initialized before first usage.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu reviewed (commented):
Looks good now, although need to introduce fs_file_t_init() usage in the same commit which introduce the fix (or before).

Also I would introduce a test which will check whether fs_open() asserts on uninitialized fs_file_t variable

pabigot reviewed (commented):
Unless I've missed something I don't see any need to change the commit order; the following command shows all commits pass all tests in `tests/subsys/fs`.  

```
git rebase -x 'twister -p k64f -T tests/subsys/fs' aed1db5782
```

Otherwise this looks good to me, though I agree a test that `fs_open` does return `-EBUSY` if the file structure is in use would be appropriate.
